[WIP/Experiment] Sing Screen redesign

Jiiniasu opened the pull request:
I've had some different ideas of "reinventing" the sing screen mainly to support more voices and to make it less disadvantageous to be the "middle" singer when playing games with 3, 5 or 6 players per screen.

After discovering Adobe XD I spent some time drawing on a concept (with all the fancy background blur I dream of having). Results below:
![singscreen](https://user-images.githubusercontent.com/1884464/44616049-0254f000-a849-11e8-92f6-96dbc07311dc.png)
![singscreen 2](https://user-images.githubusercontent.com/1884464/44616052-13056600-a849-11e8-996a-05f10ab91025.png)

So, "end goal" ideas I have (also mixed with some feature ideas...):

- Scrolling notes with a static "judgement line"
- Fixed width note segments
- MIDI song support (for songs with variable tempo and whatnot)
- Slur/note bend capabilities
- Scrolling or static lyric texts for each player
- Large tone helper displayed when microphone is used
- other things I've temporarily forgot

After a bit of head-scratching and keyboard flailing I've now reached this point:
![cur](https://user-images.githubusercontent.com/1884464/44616202-b48db700-a84b-11e8-9652-69e1f1f4f792.png)

Still missing a lot of the ideas, but I guess enough stuff is implemented for people to try out the general idea and feel.

What do you think? Are there aspects I have overlooked? Any other ideas sprouting out of this? Any questions on my thoughts?

Commit:
Fix: Fill in sung lyrics with correct colour when using Jump/Zoom

Commit:
Make it possible to generate CLyrics based on existing ones

Commit:
Pass along the AllMonitors property to the Text of CLyrics

Commit:
Expose color properties of CLyrics

Commit:
Add ability to set Alignment of CLyrics

Commit:
Expose actual width of CLyric

Commit:
Remove empty background

Commit:
First draft of new ScreenSing with individual lyrics

Commit:
Fix: Correct incorrectly set LyricStyle

Commit:
Remove unused code left from troubleshooting

derivator commented:
I think the note lines look very small, especially for the 2-player version where more space would be available. Remember that people play this standing up, from 4 meters away, on a small laptop screen.

For MIDI support I don't really see the gain, to my knowledge, most songs set their #BPM much higher than their actual tempo to gain resolution anyway, so variable tempo shouldn't be a problem. Although for the few song charts I made, my method was actually to map out the whole song as MIDI first and then use the conversion feature in UltraStar, because I could never get correct sync with any of the editors I tried. If Vocaluxe could do that, it would certainly be nice.

Another important point that applies to MIDI but also to slur/note bending is that such changes of course require new songs to be made that would be incompatible with other software using the current format. I don't know how alive the community is at this point, but even if it is, I think the willingness to make songs exclusively for Vocaluxe would be quite low? Innovation is nice of course, I just don't know if it's worth your time at this point, but maybe I'm wrong... At least in my circles, Vocaluxe has unfortunately not gotten the recognition it deserves, as the clearly superior singing game (*cough*). People still prefer UltraStar because of *one* feature missing from Vocaluxe (Team Party VS) and because it's what they know and love. Even though UltraStar is a total shit-show usability wise.

I'm not sure I'm a fan of scrolling lyrics/notes either, I feel like that works for a reaction based game like guitar hero, for singing I like taking a second to read the line and look at the shape of the melody, not anticipating new words as they rush by.

Jiiniasu commented:
@derivator, many thanks for the thorough response!

I do realise that there may be situations where one would be restricted to a small monitor at a long distance. At the same time running say, a 15.6" display at 4m isn't really optimal for anything to be honest and I think it would bring more harm than good to optimise for such setups. However I can agree that the lines get really thin in the concept (concept is with hard difficulty in mind).

The Vocaluxe setups I'm used to run have either been "small" multiprojector (~50-60" per projector) at 5 m, large single projector (100"+) at 3-8m and small screens (12.5-21.5") at 0.5-1m. So I might only cover the extremes with my experinces... It would be interesting to know what people tend to run the software on. However I believe that optimising for an old modest small TV-setup on about 32" over a couple of meters could be a good direction.

With the MIDI support I was thinking about the already existing "standard" that for example Performous uses. (I haven't thoroughly researched what it actually is, but notes and lyrics seems contained in a notes.mid.) This should prevent the creation of a Vocaluxe proprietary midi format.

When I've created songs myself I've found variable BPM to be a major annoyance to handle in the current song format. The increasing of BPM to "gain resolution" is in my opinion a very ugly "fix" which also makes it impossible to make any features that has to be tied to the "true" BPM. For example I tried to make the lyric helper flash with the beat only to give up and conclude that it won't ever work due to the way the BPM property is practically used.

Scrolling vs static notes is a difficult one. On one hand I also like seeing the "overview" of the melody, on the other hand I feel I get enough by sightreading a little bit ahead. When playing songs I haven't heard before I actually think I'd prefer scrolling notes as the entire line becomes too much to process at times. Additionally the time/tempo isn't really represented when lines of different lengths are normalised into a fixed width so I get easily thrown off if the line goes faster/slower than anticipated.

For the scrolling vs static lyrics (ie, the text) I don't have any particularily strong general preference. I can go for scrolling lyrics for most songs, but for songs I haven't heard or are very quick I think I perfer static lyrics.

I'm interested to know more on how your thoughts are about differentiating what you call "reaction based games" and the SingStar clones/variants. Personally I acknowledge that there are differences in the way of playing, but I wouldn't say there are differences in the general mechanics. I mean, you are presented with a musical note target, and you try to hit it. That's at least how I see it.

(If you made it all the way through here, thanks for reading!)

flokuep commented:
Thanks for your proposal! I really like the idea of having lyrics just below notes, that helps singing unknown songs a lot. (and the blurry background is very nice eyecandy, might be good to have this some time). Also singing duets with different lyrics for each part should be easier with this.

As @derivator mentioned, I think the note-lines and lyrics are too small. They are actually smaller than player's name and lyrics and note-lines are the more important information. (side question: what's the difficulty level in your screenshots?)

I currently don't know if I'll like the scrolling-notes/lyrics. It might be harder to play, because there is so much moving on the screen, might need to try it.
> Additionally the time/tempo isn't really represented when lines of different lengths are normalised into a fixed width so I get easily thrown off if the line goes faster/slower than anticipated.

That's a really good point! 

Regarding BPM/MIDI: BPM is a more technical value, that simply seperates a minute into beats, and notes could only be set to this beats. So the higher the value, the more accurate the notes can be. If you have a song with different tempo, you could simply use the least common multiple, I guess. Don't know what this mid-file contains - Do you have a sample one?

Jiiniasu commented:
@flokuep, thanks for your input as well!

The concept images had the hard difficulty in mind (to plot out the "worst case scenario" of thin lines).
I should've maybe taken a real screenshot with all three difficulties present, so I've done so below:
![2018-08-27](https://user-images.githubusercontent.com/1884464/44671663-74cef700-aa26-11e8-9a93-1e21c20baefc.png)
(Not sure wth is up with two of the notes on hard that somehow are thinner than the rest, possibly rounding related...)

To be honest I'm not sure how one is supposed to get the notes larger than they already are. As you see in the Guest Easy players there are tiny notes that are just slightly too small to fit the start and end of a note-line. If I would increase the height, its relative width would decrease even more. I would then need to at least equally scale the width and the height to prevent short notes to be invisible or close to invisible. One could argue that the width can be increased a bit more but I believe it would have negative impact on the overall look as it would become very cramped. I'd gladly receive any good ideas on this. Scrolling notes however, would solve the width issue as a note-section width would be constant.

For the BPM/Tempo thing; I strongly disagree with it being solely a measurement of accuracy, even though in this case it's usually used as such. For variable tempo one would reach ridiculous BPM values if you are unlucky. I can't remember which song I was struggling with to get good results that had variable tempo, but I remember that the BPM would have needed to be somewhere around 600 to fit the different tempos. 

Anyway, I think it is important to use the musical definition of BPM in a music game. This accuracy thing reminds me of another issue with the current format - triplets. That would also drive the BPM values to the skies to "get right". I imagine many nice QoL things that can be added if only the BPM was used as BPM and not an abstract "accuracy value". Beat countdown, graphic elements animated in tempo, possibility to build visualisers based on the correct BPM etc.

The midi format I believe is either the same as Rock Band used or maybe a derivative of it, so it would contain tracks for guitar, bass/rythm, drums, vocals + harmonies etc. and each track would have notes + marker notes for different features. [Example notes.mid](https://unsuspicious.website/midiexample/notes.mid)

derivator commented:
On the BPM discussion I would like to add that some songs that were recorded in pre-DAW/quantaize-everything times (or live, or by hipsters/luddites) simply don't have a consistent tempo, the tempo is sometimes 120BPM, sometimes 118 and sometimes 125, whatever the band felt like in that moment. And this is more common than one would think. [This discussion](https://github.com/UltraStar-Deluxe/USDX/issues/349) also touches on some points of why the BPM system is the way it is.

flokuep commented:
> To be honest I'm not sure how one is supposed to get the notes larger than they already are.

Yeah, right. But we might think about the lyrics... What about having them just below the belonging notes? Don't know if this would look silly or hard to read, but there is _enough_ space...

> Anyway, I think it is important to use the musical definition of BPM in a music game. This accuracy thing reminds me of another issue with the current format - triplets. That would also drive the BPM values to the skies to "get right". I imagine many nice QoL things that can be added if only the BPM was used as BPM and not an abstract "accuracy value". Beat countdown, graphic elements animated in tempo, possibility to build visualisers based on the correct BPM etc.

You're right regarding the _meaning_ of the BPM value, but I'm not sure if we could change a lot here... We can't simply break the common song file format.

Commit:
Remove unused variable

Commit:
Encapsulate allmonitors in a public property

Commit:
Center Note Lines

Commit:
Shift Notes in Lines to Center

Commit:
Revisit Screen Sing Design

Commit:
First Implementation of Scrolling Notes

Also FreeStyle Notes

Commit:
Merge remote-tracking branch 'refs/remotes/Vocaluxe/develop' into lyrics

Commit:
Expose CurrentBeatF

Commit:
Have the notes scroll with the current beat

Commit:
Enlarge tone helper

Commit:
Remove commented out code

Commit:
Make _JudgementLine readonly

Commit:
Do something with AllMonitors to please SonarCloud

Commit:
Little cleaning

Commit:
Increase min note size

To prevent easy notes to become squashed if they are short

Commit:
Remove unused variable

Jiiniasu commented:
I could have sworn that I had written a reply for the last two comments... Anyway...
Thanks @derivator for linking the discussion regarding BPM, it was an interesting read!

Been playing around and doing some adjustments and managed to get some kind of scrolling notes working. All fancy effects like flares and golden notes are broken as of now, but the general gameplay feel is "testable".

The note-area height has been increased and the font size of the lyrics have been slightly increased as well.

![vocaluxe blue sunrise 0 4 0 x64 0 4 0-343-g00023eb 2018-09-30 09_38_32](https://user-images.githubusercontent.com/1884464/46255762-ad3c7780-c4a1-11e8-80da-23398a4a6e2b.png)

WDYT?

Commit:
Correct rating bar graphics and positions

Commit:
Add _NoteWidth

Commit:
Expose GetCurrentBeat()

Commit:
Expose ToneValid()

Commit:
Merge remote-tracking branch 'refs/remotes/Vocaluxe/develop' into lyrics

Commit:
Correct Tone Helper placement

Commit:
Tone range compression and more

First dirty implementation of tone range compression for songs spanning
past two octaves
Breaking out some code etc.
In need of cleaning and testing...

Commit:
Remove unused parameter

Commit:
use .Any() for empty-test.

Commit:
create _DrawNotes()

Commit:
Change in note drawing functions

Commit:
change x + y to right

Commit:
Correct _genoterect in golden notes

hendrikbl commented:
To be honest: I don't really like the redesign 🙈
I appreciate the work you put in this and I like the looks of it in general (especially the blurry background looks nice!) but I don't like it in a singing game like vocaluxe. It takes away so much of the video and makes notes and lyrics smaller without any real advantage over the old design. Especially when it comes to two players.

Jiiniasu commented:
@hendrikbl,
Thanks for the feedback!

On the topic of the notes/lyrics; what are your opinions regarding scrolling notes?

Jiiniasu commented:
@hendrikbl,
Yes, this has got an early attempt of scrolling notes implemented.

Stefan1200-de commented:
Thanks for sharing your work with us. Now I tested this build too.
I like that design in general, it looks beautiful (really!). But like hendrikbl, I don't see real advantages for Vocaluxe in it. Text size is smaller and why displaying the same text for every player again? Also more parts of the video will be hidden (even if a singer has nothing to sing).
Of course, this is just my opinion.

Jiiniasu commented:
> why displaying the same text for every player again?

Well, it removes the disadvantage of being a player in the middle of the screen if you show the same text. However all players having the same text might not always be the case. If you have multiple voices this will let us use one voice per player if one wants to, or mix and match without having the lyrics too far away.
See example below from a three voice test of the song Bohemian Rhapsody by Queen:
![Screenshot_5](https://user-images.githubusercontent.com/1884464/72293451-d7448980-3653-11ea-9995-ae416b3fd616.png)

> Also more parts of the video will be hidden

Yes, that's one disadvantage of the design, but at the same time it is a sorry attempt to get some kind of consistant level of contrast for the note bars.

Stefan1200-de commented:
Oh, I never had files with more than two voices. Of course, with three voices that change works better than current solution. Is it possible to hide the players for the time they have nothing to sing? I don't know if that looks good, but maybe.

Jiiniasu commented:
> Oh, I never had files with more than two voices.

That's likely because not even vocaluxe can handle more than two voices "out of the box". I believe most of the UltraStar-derived games only handle duets at most.

>Is it possible to hide the players for the time they have nothing to sing?

Sure thing, it's already built in, I personally have it turned off, but the "Fade player info" setting in the options should do that. Not sure how much I tested it tbh, so that might need some tweaking...

Stefan1200-de commented:
> Sure thing, it's already built in, I personally have it turned off, but the "Fade player info" setting in the options should do that. Not sure how much I tested it tbh, so that might need some tweaking...

Found it, tested it, works. Thanks for the hint.

I realized, that the twinkle at the golden notes will be displayed wrong the most of the times. Do you know that already?

Jiiniasu commented:
> I realized, that the twinkle at the golden notes will be displayed wrong the most of the times. Do you know that already?

Well, yes. I was not able to figure out how to properly solve that one back when I did the bulk of the code... The same goes for the sparkling effect for the golden note bars, which is why they are missing.
